### PR TITLE
feat(broker): BPMN merging parallel gateway

### LINKED ISSUE #1231: I can run a workflow with a merging parallel gateway

### PR DESCRIPTION
**Pull request by ThorbenLindhauer**
- refactors the way we count tokens, such that the counting happens in a
  single place and is transparently for the step handlers; this will
  hopefully ease understanding of the overall behavior
- adds concept to defer events by storing them in the index, which
  is used when a sequence flow is signalled that does not immediately
  trigger the gateway
- there is optimation potential for the implementation of the merge and
  the index data structure. I chose the simplest approach for both,
  to be flexible when we migrate the index to rocks db and to avoid
  effort before we are even aware of performance issues

closes #1231

### CODE REVIEW

File: broker-core/src/main/java/io/zeebe/broker/exporter/stream/ExporterStreamProcessorState.java [hunk 1]
```diff
@@ -95,7 +95,7 @@ public long getPosition(final byte[] exporterId) {
   public ExporterRecord newExporterRecord() {
     final ExporterRecord record = new ExporterRecord();
 
-    try (final RocksIterator iterator = getDb().newIterator()) {
+    try (RocksIterator iterator = getDb().newIterator()) {
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/exporter/stream/ExporterStreamProcessorState.java]: always fun the auto formating :sweat_smile: or do you removed the final by yourself ? :)

File: broker-core/src/main/java/io/zeebe/broker/workflow/index/IndexedRecord.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    this.state = state;
+  }
+
+  public WorkflowInstanceRecord getValue() {
+    return value;
+  }
+
+  public void setValue(WorkflowInstanceRecord value) {
+    final int encodedLength = value.getLength();
+    valueBuffer.checkLimit(encodedLength);
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/index/IndexedRecord.java]: What is the reason for this?

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/sequenceflow/ParallelMergeHandler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    final List<IndexedRecord> mergeableRecords = getMergeableRecords(gateway, scopeInstance);
+
+    if (mergeableRecords.size() == gateway.getIncoming().size()) {
+
+      mergeableRecords.forEach(
+          r -> eventOutput.consumeDeferredEvent(scopeInstance.getKey(), r.getKey()));
+
+      final WorkflowInstanceRecord value = context.getValue();
+      value.setActivityId(gateway.getId());
+      context.getOutput().writeNewEvent(WorkflowInstanceIntent.GATEWAY_ACTIVATED, value);
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/sequenceflow/ParallelMergeHandler.java]: Do we merge the payload as well ? When Yes, where? :D

File: broker-core/src/main/java/io/zeebe/broker/workflow/processor/gateway/ExclusiveSplitHandler.java [hunk 4]
```diff
@@ -43,16 +43,31 @@ public void handle(BpmnStepContext<ExecutableExclusiveGateway> context) {
...
-      context.raiseIncident(ErrorType.CONDITION_ERROR, e.getMessage());
+      raiseIncident(context, e.getMessage());
     }
   }
 
+  private void raiseIncident(
+      BpmnStepContext<ExecutableExclusiveGateway> context, final String errorMessage) {
+    context.raiseIncident(ErrorType.CONDITION_ERROR, errorMessage);
+
+    // TODO: this is a hack to avoid that we believe this token is consumed when the incident is
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/workflow/processor/gateway/ExclusiveSplitHandler.java]: This a problem when we decrement on event consuming and increment, if events are written right?
Is it possible that we only increment on scope creation and decrement on scope termination? I'm not sure if we already discussed that sorry :sweat_smile: Maybe we can discuss this f2f again.

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

**Review by ChrisKujawa — APPROVED** [review]

**Comment by ThorbenLindhauer** [comment]
bors retry

**Comment by ThorbenLindhauer** [comment]
bors retry

**Comment by menski** [comment]
bors retry

**Pull request merged**